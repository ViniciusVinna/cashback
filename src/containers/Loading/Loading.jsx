import React, { useEffect, useRef } from 'react';
import ReactVivus from 'react-vivus';
import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';
import { useSelector } from 'react-redux';

import defaultImage from 'assets/brands/boticario-light.svg';

import { LoadingStyled } from './Loading.styles';

const { LoadingWrapper, Status } = LoadingStyled;

const Loading = () => {
  const { isLoading, isScrollLocked, statusMessage } = useSelector(state => state.app);
  const loadingRef = useRef(null);

  useEffect(() => {
    if (isScrollLocked) {
      disableBodyScroll(loadingRef);
    }

    return () => {
      enableBodyScroll(loadingRef);
    };
  }, [isScrollLocked]);

  const handleAnimation = (myVivus) => {
    myVivus.play(myVivus.getStatus() === 'end' ? -1 : 1);
  };

  /* instanbul ignore else */
  if (!isLoading) {
    return null;
  }

  return (
    <LoadingWrapper
      ref={loadingRef}
      data-testid="Loading"
    >
      <ReactVivus
        id="loading"
        option={{
          file: defaultImage,
          type: 'scenario',
          animTimingFunction: 'EASE_OUT_BOUNCE',
          duration: 50,
        }}
        style={{ height: 'auto', width: '100%' }}
        callback={handleAnimation}
      />

      {statusMessage && (
        <Status>{statusMessage}</Status>
      )}
    </LoadingWrapper>
  );
};

export default Loading;
