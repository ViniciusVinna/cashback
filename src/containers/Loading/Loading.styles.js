import styled from '@emotion/styled';
import { css } from '@emotion/core';

export const LoadingStyled = {
  LoadingWrapper: styled('div')(({ theme, isLoading }) => css`
    align-items: center;
    backface-visibility: hidden;
    background-color: ${theme.color.n100};
    bottom: 0;
    display: flex;
    height: 100vh;
    justify-content: center;
    left: 0;
    opacity: ${isLoading ? '0.9' : '0'};
    padding: ${theme.space.xl};
    pointer-events: ${isLoading ? 'none' : 'all'};
    position: fixed;
    right: 0;
    top: 0;
    transition: ${theme.transition.bezier};
    width: 100%;
    z-index: ${isLoading ? theme.zIndex.loading : theme.zIndex.deep};

    & #loading {
      width: 100%;
      max-width: 300px;
    }
  `),
};
