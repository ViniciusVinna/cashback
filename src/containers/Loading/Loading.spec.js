import React from 'react';
import { useSelector } from 'react-redux';

import Loading from './Loading';

let mockStore = {
  app: {
    isLoading: false,
    isScrollLocked: false,
    statusMessage: false,
  },
};

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
}));

describe('Loading', () => {
  beforeEach(() => {
    useSelector.mockImplementation(callback => callback(mockStore));
  });

  afterEach(() => {
    useSelector.mockClear();
  });

  it('it should not render Loading', () => {
    const { container } = render(<Loading />);
    expect(container).toMatchSnapshot();
  });

  it('it should render Loading when isLoading is defined', () => {
    mockStore = {
      app: {
        isLoading: true,
        isScrollLocked: true,
        statusMessage: 'Carregando dados',
      },
    };

    const { container } = render(<Loading />);
    expect(container).toMatchSnapshot();
  });
});
