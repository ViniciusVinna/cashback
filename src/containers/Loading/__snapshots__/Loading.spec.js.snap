// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading it should render properly 1`] = `<div />`;
